designs/SolutionLoad-CsprojPackageReferenceRestore.md

```diff
@@ -0,0 +1,110 @@
+
+Now you're likely asking yourself, how do we have 3 restores?
+
+Even if NuGet doesn't expect a nomination from certain projects, it will *still* batch the restores ran for that nomination call because they might've happened before the last nomination call for an SDK based project.
+
+Given that VS loads projects in the order in which they are defined in the project file, it happens almost consistently that the nomination calls from those 2 VSIX projects come after the last SDK based project nomination.
+
+While, there's many csproj.dll projects supporting PackageReference, we have so far identified that Xamarin and VSIX projects are the ones that do call nomination on solution load.
+
+### What are the consequences
+
+Talk about the time wasted.
+
+## Who are the customers
+
+Customers with csproj.dll based PackageReference projects.
+
+## Goals
+
+* Propose a coordination mechanism to avoid extra restores.
+* Given that this is likely to be a multi-component, track those project-systems and drive the changes to adopt this new mechanism.
+
+## Non-Goals
+
+## Solution
+
+What makes solution load running predictable in SDK based projects situation is that NuGet knows what projects are supposed to nominate. As described above, NuGet can't even restore an SDK based project if the nomination has not come in, because the data is pushed to NuGet.
+
+The proposal is to extend this `knowledge` to NuGet in csproj.dll projects.
+In particular, we are proposing that NuGet tracks the `nominate` call for those projects and waits for all the nominations similarly to how we wait for SDK based projects.
+The reference data will still be read through the poll model. The reason for this is if those projects could move to new model, they would've already. We are looking for a minimal investment approach here.
+
+### How do projects pre-register
+
+Project-systems have capability model. Whenever NuGet detects a project that's csproj.dll-based PackageReference, it can check for an additional capability that will indicate that a project is expected to call nominate at solution load.
+
+### What capability in particular
+
+We *might* be able to re-use some existing capabilities, but we need to be able to correctly identify and bucketize the right groups of project.
+The particular capability name is up for discussion at this point.
+
+[See Open Questions](#open-questions)
+
+## Future Work
+
+* Once the design has been agreed on, this will require changes in potentially multiple project-systems. The changes should be minimal and low effort.
+
+## Open Questions
+
+* What should be the capability name?
+
+`PackageReferences` is a capability we can consider. I am not confident whether `PackageReference` suggests a communication method with NuGet (ie. push vs pull for getting the references). Is redefining a meaning of a capability acceptable?
+
+## Considerations
+
+No other solutions for batching were considered.

```

Comment:
What about ignoring nominations for project types we don't expect nominations from?  Although the XMLDoc for the nominate APIs explicitly says that the return task will be complete when the restore is complete. Therefore it's hard for us to claim that nominations aren't supposed to be a restore.

Reply:
The problem is that https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.SolutionRestoreManager.Interop/IVsSolutionRestoreService2.cs was added specifically with that intention. 

Even if the completion means `project is ready` we can't guarantee that because it's hard to differentiate between solution load and non-solution load scenarios.